### PR TITLE
Correctly set payment information when using paypal

### PR DESCRIPTION
**Pull request by therool**
### Description
Correctly set payment method information when using PayPal Express so the checkout agreements data is correctly added to the request and validated correctly. 
Edited the  `Magento_Paypal
/js/action/set-payment-method` to use the `Magento_Checkout/js/action/set-payment-information` so the additional payment/checkout data is correctly added.

### Fixed Issues (if relevant)

1. magento/magento2#11885: Paypal Can't Accept Checkout Agreements Before Routing to PayPal

### Manual testing scenarios

1. Create required terms and conditions
2. Enable paypal express checkout
3. Checkout with paypal express in the frontend - does not work because terms and coniditons data is not correctly added to the set payment method information request

### Contribution checklist
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All new or changed code is covered with unit/integration tests (if applicable)
 - [x] All automated tests passed successfully (all builds on Travis CI are green)

### CODE REVIEW

File: app/code/Magento/Paypal/view/frontend/web/js/action/set-payment-method.js [hunk 1]
```diff
@@ -10,42 +10,12 @@ define([
     'mage/storage',
     'Magento_Checkout/js/model/error-processor',
```
Comment (YevSent) [inline app/code/Magento/Paypal/view/frontend/web/js/action/set-payment-method.js]: Please, remove unused dependencies

File: app/code/Magento/Paypal/Test/Unit/Model/ExpressTest.php [hunk 2]
```diff
@@ -161,12 +161,21 @@ public function testAssignData()
...
+        );
+
         $data = new DataObject(
             [
                 PaymentInterface::KEY_ADDITIONAL_DATA => [
                     Express\Checkout::PAYMENT_INFO_TRANSPORT_BILLING_AGREEMENT => $transportValue,
                     Express\Checkout::PAYMENT_INFO_TRANSPORT_PAYER_ID => $transportValue,
-                    Express\Checkout::PAYMENT_INFO_TRANSPORT_TOKEN => $transportValue
+                    Express\Checkout::PAYMENT_INFO_TRANSPORT_TOKEN => $transportValue,
+                    \Magento\Framework\Api\ExtensibleDataInterface::EXTENSION_ATTRIBUTES_KEY => $extensionAttributeMock
```
Comment (YevSent) [inline app/code/Magento/Paypal/Test/Unit/Model/ExpressTest.php]: These changes do not cover fixed bug. Please, cover them by the unit test.

**Review by YevSent — CHANGES_REQUESTED** [review]

**Comment by ishakhsuvarov** [comment]
Hi @therool
Could you please check failing JS Unit so we could proceed with the merge?
Thank you

**Comment by therool** [comment]
@ishakhsuvarov I will, I just will fix the commits, I pushed something unnecessary

**Comment by therool** [comment]
@ishakhsuvarov fixed. I will try to add one more JS unit test later `set-payment-information-mixin.js.test` under `dev/tests/js/jasmine/tests/app/code/Magento/CheckoutAgreements/frontend/js/model/set-payment-information-mixin.js.test` which will test that the agreement ids are added as extension attributes to the payment data when the `Magento_Checkout/js/action/set-payment-information` is being called.

**Comment by ishakhsuvarov** [comment]
@therool Thanks. Should we wait for the test then?
Please let us know when we can continue with review & testing.

**Comment by ishakhsuvarov** [comment]
Hi @therool 
Do you still want to continue work on this PR? Any help needed from us?

**Comment by therool** [comment]
@ishakhsuvarov I will continue to work on this as everything is done I just want to add jasmine test for the checkout agreement mixins whcih is not necessary.

Maybe you know how to correctly test mixins with jasmine and squire.js as for mixins I need the squire to include the logic from https://github.com/magento/magento2/blob/2.2-develop/lib/web/mage/requirejs/mixins.js#L5 but I can't include `"mage/requirejs/mixins"` in the dependencies as that breaks all others tests. The test would look something like this:
```js
/**
 * Copyright © Magento, Inc. All rights reserved.
 * See COPYING.txt for license details.
 */

define([
    'squire'
], function (Squire) {
    'use strict';

    describe('Magento_CheckoutAgreements/js/model/place-order-mixin', function () {
        var injector = new Squire(),
            mocks = {
                'Magento_CheckoutAgreements/js/model/agreements-assigner': jasmine.createSpy('agreements-assigner'),
		'Magento_Checkout/js/action/place-order': jasmine.createSpy('placeOrder'),
            },
            placeOrderAction;

        beforeAll(function (done) {
            injector.mock(mocks);
            injector.require(['mixins!Magento_Checkout/js/action/place-order'], function (placeOrder) {
                placeOrderAction = placeOrder;
                done();
            });
        });

        it('Magento_CheckoutAgreements/js/model/agreements-assigner is called', function () {
            var messageContainer = jasmine.createSpy('messageContainer'),
                paymentData = {};
            placeOrderAction(paymentData, messageContainer);
            expect(mocks['Magento_CheckoutAgreements/js/model/agreements-assigner']).toHaveBeenCalled();
            expect(mocks['Magento_Checkout/js/action/place-order']).toHaveBeenCalled();
        });
    });
});
```

And this is the one I currently have written but It does not work correctly because I can't get the ` 'mixins': require('mixins')` working.
```js
/**
 * Copyright © Magento, Inc. All rights reserved.
 * See COPYING.txt for license details.
 */

define([
    'squire',
    'jquery',
    'mage/requirejs/mixins'
], function (Squire, $) {
    'use strict';

    describe('Magento_CheckoutAgreements/js/model/set-payment-information-mixin', function () {
        var injector = new Squire(),
            mocks = {
                'Magento_Checkout/js/model/quote': jasmine.createSpyObj('quote', ['getQuoteId', 'billingAddress']),
                'Magento_Checkout/js/model/url-builder': jasmine.createSpyObj('urlBuilder', ['createUrl']),
                'mage/storage': jasmine.createSpyObj('storage', ['post']),
                'Magento_Checkout/js/model/error-processor': jasmine.createSpy('errorProcessor'),
                'Magento_Customer/js/model/customer': jasmine.createSpyObj('customer', ['isLoggedIn']),
                'Magento_Checkout/js/action/get-totals': jasmine.createSpy('getTotalsAction'),
                'Magento_Checkout/js/model/full-screen-loader': jasmine.createSpyObj('fullScreenLoader', ['startLoader']),
                'Magento_CheckoutAgreements/js/model/agreements-assigner': jasmine.createSpy('agreements-assigner'),
                'mixins': require('mixins')
            },
            setPaymentInformationAction;

        beforeAll(function (done) {
            window.checkoutConfig = {
                checkoutAgreements: {
                    isEnabled: true
                }
            };
            injector.mock(mocks);
            injector.require(['mixins!Magento_Checkout/js/action/set-payment-information'], function (setPaymentInformation) {
                setPaymentInformationAction = setPaymentInformation;
                done();
            });
        });

        it('Magento_CheckoutAgreements/js/model/agreements-assigner is called', function () {
            mocks['mage/storage'].post.and.callFake(function () {
                return $.Deferred();
            });
            var messageContainer = jasmine.createSpy('messageContainer'),
                paymentData = {};
            setPaymentInformationAction(messageContainer, paymentData);
            expect(mocks['Magento_CheckoutAgreements/js/model/agreements-assigner']).toHaveBeenCalled();
            expect(mocks['mage/storage'].post).toHaveBeenCalled();
        });
    });
});
```

**Comment by omiroshnichenko** [comment]
Hi, @therool. Here is a sample how you can test mixin:
```js
define(['squire'], function (Squire) {
    'use strict';

    var injector1 = new Squire(),
        injector2 = new Squire(),
        mixin, paymentInformation, mock1, mock2;

    beforeEach(function (done) {
        mock1 = {
        };
        mock2 = {
            'Magento_Checkout/js/model/quote': jasmine.createSpyObj(
                'quote',
                ['totals', 'getQuoteId', 'billingAddress']
            ),
            'Magento_Checkout/js/action/get-totals': jasmine.createSpy()
        };
        window.checkoutConfig = {
            checkoutAgreements: {
                isEnabled: 1
            }
        };
        injector1.mock(mock1);
        injector2.mock(mock2);
        injector1.require(['Magento_CheckoutAgreements/js/model/set-payment-information-mixin'], function (Mixin) {
            mixin = Mixin;
            injector2.require(['Magento_Checkout/js/action/set-payment-information'], function (PaymentInformation) {
                paymentInformation = PaymentInformation;
                done();
            });
        });
    });

    describe('Magento_CheckoutAgreements/js/model/agreement-validator', function () {
        describe('"validate" method', function () {
            it('Check with non existing checkboxes', function () {
                mixin(paymentInformation)('', {});
                expect(true).toBe(true);
            });
        });
    });
});
```

**Comment by therool** [comment]
Hi @omiroshnichenko, thanks for the sample, I added mixin tests so everything should be fine now.

**Comment by AppSimone** [comment]
any news about dank00 patch, I've tried to install but it fails.

  - Applying patches for magento/module-paypal (1)
    ~ space48/magento2-patches: patches/Magento_Paypal/paypal-express-checkout-agreements.patch [NEW]
      Fix Paypal Checkout Agreements https://github.com/magento/magento2/issues/11885

**Pull request merged**